File: 2287_rearrange-characters-to-make-target-string/csharp/solution.cs
public class Solution
{
    public int RearrangeCharacters(string s, string target)
    {
        // var sCount = s.GroupBy(c => c).ToDictionary(g => g.Key, g => g.Count());
        // var targetCount = target.GroupBy(c => c).ToDictionary(g => g.Key, g => g.Count());
        //
        // var result = int.MaxValue;
        //
        // foreach (var (character, count) in targetCount)
        // {
        //     if (!sCount.ContainsKey(character))
        //         return 0;
        //     result = Math.Min(result, sCount[character] / count);
        // }
        //
        // return result;

        const int MAX_LETTERS = 26;

        int[] sCount = new int[MAX_LETTERS];
        int[] targetCount = new int[MAX_LETTERS];

        foreach (char c in s)
            sCount[c - 'a']++;
        foreach (char c in target)
            targetCount[c - 'a']++;

        int result = int.MaxValue;

        for (int i = 0; i < MAX_LETTERS; i++)
        {
            if (targetCount[i] > 0)
            {
                result = Math.Min(result, sCount[i] / targetCount[i]);
            }
        }

        return result;
    }
}

File: 2287_rearrange-characters-to-make-target-string/rust/solution.rs
use std::cmp;

#[cfg(feature = "harness")]
pub struct Solution;

impl Solution {
    pub fn rearrange_characters(s: String, target: String) -> i32 {
        let max_letters = 26;

        let mut s_count = vec![0; max_letters];
        let mut target_count = vec![0; max_letters];

        for c in s.chars() {
            s_count[c as usize - 'a' as usize] += 1;
        }

        for c in target.chars() {
            target_count[c as usize - 'a' as usize] += 1;
        }

        let mut result = i32::MAX;

        for i in 0..max_letters {
            if target_count[i] > 0 {
                result = cmp::min(result, s_count[i] / target_count[i]);
            }
        }

        result
    }
}

